From 2c67ec1b04b7d7d7b64e3fee9acea7ea966ad9d8 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Sun, 17 Feb 2013 23:15:32 -0600
Subject: [PATCH] Fire PlayerInteractEvent when anyone steps on a pressure
 plate

In order to not spam the PlayerInteractEvent, we need to keep track of wether or not a plugin allows/denys a player.

This does modify PlayerInteractEvent firing, but it shouldn't be 'breaking'.

diff --git a/src/main/java/net/minecraft/server/BlockPressurePlateBinary.java b/src/main/java/net/minecraft/server/BlockPressurePlateBinary.java
index 25ef883..f67ad3c 100644
--- a/src/main/java/net/minecraft/server/BlockPressurePlateBinary.java
+++ b/src/main/java/net/minecraft/server/BlockPressurePlateBinary.java
@@ -8,6 +8,7 @@ import org.bukkit.event.entity.EntityInteractEvent; // CraftBukkit
 public class BlockPressurePlateBinary extends BlockPressurePlateAbstract {
 
     private EnumMobType a;
+    public java.util.Map<Object, Boolean> entities = new java.util.HashMap<Object, Boolean>(); // CraftBukkit
 
     protected BlockPressurePlateBinary(String s, Material material, EnumMobType enummobtype) {
         super(s, material);
@@ -37,6 +38,19 @@ public class BlockPressurePlateBinary extends BlockPressurePlateAbstract {
             list = world.a(EntityHuman.class, this.a(i, j, k));
         }
 
+        // CraftBukkit start
+        if (!this.entities.isEmpty()) {
+            Iterator iterator = this.entities.keySet().iterator();
+
+            while (iterator.hasNext()) {
+                Entity entity = (Entity) iterator.next();
+                if (!list.contains(entity)) {
+                    iterator.remove();
+                }
+            }
+        }
+        // CraftBukkit end
+
         if (list != null && !list.isEmpty()) {
             Iterator iterator = list.iterator();
 
@@ -44,11 +58,21 @@ public class BlockPressurePlateBinary extends BlockPressurePlateAbstract {
                 Entity entity = (Entity) iterator.next();
 
                 // CraftBukkit start - Call interact event when turning on a pressure plate
-                if (this.c(world.getData(i, j, k)) == 0) {
+                Boolean result = this.entities.get(entity);
+
+                if (true || this.c(world.getData(i, j, k)) == 0) {
                     org.bukkit.World bworld = world.getWorld();
                     org.bukkit.plugin.PluginManager manager = world.getServer().getPluginManager();
                     org.bukkit.event.Cancellable cancellable;
 
+                    if (entity == null) {
+                        continue;
+                    }
+
+                    if (!entity.az() && result != null) {
+                        return result ? 15 : 0;
+                    }
+
                     if (entity instanceof EntityHuman) {
                         cancellable = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerInteractEvent((EntityHuman) entity, org.bukkit.event.block.Action.PHYSICAL, i, j, k, -1, null);
                     } else {
@@ -61,9 +85,9 @@ public class BlockPressurePlateBinary extends BlockPressurePlateAbstract {
                         continue;
                     }
                 }
-                // CraftBukkit end
 
-                if (!entity.az()) {
+                if (!entity.az() && (result == null || result)) {
+                    // CraftBukkit end
                     return 15;
                 }
             }
-- 
1.8.5.2 (Apple Git-48)

